lit_parrot/model.py

```diff
@@ -44,17 +49,57 @@ def _init_weights(self, module: nn.Module) -> None:
             # https://huggingface.co/stabilityai/stablelm-base-alpha-3b/blob/main/config.json#L12
             module.eps = 1e-5
 
-    def forward(self, idx: torch.Tensor) -> torch.Tensor:
-        _, t = idx.size()
-        assert (
-            t <= self.config.block_size
-        ), f"Cannot forward sequence of length {t}, block size is only {self.config.block_size}"
+    def forward(
+        self, idx: torch.Tensor, max_seq_length: Optional[int] = None, input_pos: Optional[torch.Tensor] = None
+    ) -> Union[torch.Tensor, Tuple[torch.Tensor, List[KVCache]]]:
+        B, T = idx.size()
+
+        block_size = self.config.block_size
+        if max_seq_length is None:
+            max_seq_length = block_size
+        assert T <= max_seq_length, f"Cannot forward sequence of length {T}, max seq length is only {max_seq_length}"
+        assert max_seq_length <= block_size, f"Cannot attend to {max_seq_length}, block size is only {block_size}"
+        assert T <= block_size, f"Cannot forward sequence of length {T}, block size is only {block_size}"
+
+        if self.rope_cache is None:
+            self.rope_cache = self.build_rope_cache(idx)
+        if self.mask_cache is None:
+            self.mask_cache = self.build_mask_cache(idx)
+
+        cos, sin = self.rope_cache
+        if input_pos is not None:
+            cos = cos.index_select(0, input_pos)
+            sin = sin.index_select(0, input_pos)
+            mask = self.mask_cache.index_select(2, input_pos)
+            mask = mask[:, :, :max_seq_length, :max_seq_length]
```

Comment:
Shouldn't this be:
`mask = mask[:, :, :, :max_seq_length]`
following the changes in https://github.com/Lightning-AI/lit-llama/compare/main...Liyang90:lit-llama:cached_kvs

Reply:
Thanks for the review. Following up on your question, the code currently looks like:
```
...
            mask = mask[:, :, :, :max_seq_length]
        else:
            cos = cos[:T]
            sin = sin[:T]
            mask = self.mask_cache[:, :, :T, :T]
...
```
Like you mentioned, the changes in https://github.com/Lightning-AI/lit-llama/compare/main...Liyang90:lit-llama:cached_kvs only trimmed the last dimension in `mask`. What is the logic behind trimming only the last dimension in `mask` when using `input_pos` and trimming the last two dimensions when `input_pos == None`?

Reply:
because `mask = self.mask_cache.index_select(2, input_pos)` already selects the proper rows on dim 2.

Reply:
Fixed